### PR TITLE
multi-language installer using php arrays for definitions

### PR DESCRIPTION
**Pull request by zcwilt**
See issue #3018

### CODE REVIEW

File: zc_install/includes/languages/en_us/main.php [hunk 1]
```diff
@@ -0,0 +1,1066 @@
...
+'TEXT_COMPLETION_INSTALLATION_DIRECTORY_EXPLANATION' => 'To prevent unauthorized persons from overwriting your installation, administering your installation will be disabled until the folder has been deleted.',
+'TEXT_COMPLETION_CATALOG_LINK_TEXT' => 'Your Storefront',
+'TEXT_COMPLETION_ADMIN_LINK_TEXT' => 'Your Admin Dashboard',
+'TEXT_PAGE_HEADING_DATABASE_UPGRADE' => 'Database Upgrade',
+'TEXT_DATABASE_UPGRADE_HEADER_MAIN' => '',
+'TEXT_DATABASE_UPGRADE_STEPS_DETECTED' => 'The following list shows the various upgrade steps we detected are required for your database.',
+'TEXT_DATABASE_UPGRADE_LEGEND_UPGRADE_STEPS' => 'Please confirm your desired upgrade steps',
+'TEXT_DATABASE_UPGRADE_ADMIN_CREDENTIALS' => 'Admin Credentials (SuperUser)',
+'TEXT_VALIDATION_ADMIN_CREDENTIALS' => 'To authorize the database upgrade, you must enter an admin username and password with SuperUser permissions in your store.',
+'TEXT_HELP_TITLE_UPGRADEADMINNAME' => TEXT_DATABASE_UPGRADE_ADMIN_CREDENTIALS,
```
Comment (drbyte) [inline zc_install/includes/languages/en_us/main.php]: `TEXT_DATABASE_UPGRADE_ADMIN_CREDENTIALS` refers to `'TEXT_DATABASE_UPGRADE_ADMIN_CREDENTIALS' => 'Admin Credentials (SuperUser)',` a couple lines above, but that constant won't be defined anymore!

**Comment by drbyte** [comment]
👍

**Comment by torvista** [comment]
Would I be right in assuming that languages_installed.php needs the available languages to be listed in the array to allow selection...
i.e. a language pack cannot be dropped into the vanilla fileset and used with the installer without modifying that file?

**Comment by zcwilt** [comment]
@torvista 
yes, at the moment. The initial idea was that at least for the installer that language packs would be distributed with the release. i.e. no drop ins for installer.

I had originally played around using separate manifest files for each language pack, which would allow for dropping in language packs.

**Comment by torvista** [comment]
Without knowing what I am talking about, I would have thought that it could have auto detected what folders (en_us, es_es) were in that directory and just include them?

**Comment by zcwilt** [comment]
The languages_installed.php  contains extra meta information, mainly for building the language drop down select. 

e.g.  the displayName part

**Comment by torvista** [comment]
Is there some reason why the html formatting (spaces and line feeds) are preserved in the main file?
Can this not be compressed?
eg:

`TEXT_HELP_CONTENT_HTACCESSSUPPORT' => 'There appears to be a problem with support for ".htaccess" files.<br>Sensitive files and folders on your site, that should normally be blocked by security rules in the built-in ".htaccess" files that come with Zen Cart, are currently accessible.
	<br><br>

	Possible causes: 

	<ul style="list-style-type:square">

		<li>

			You may not be using Apache as your Web Server (".htaccess" files are unique to the Apache Web Server), or,

		</li>

		<li>

			Support for ".htaccess" is disabled or misconfigured, or,

		</li>

		<li>

			The ".htaccess" files that come with Zen Cart have not been uploaded to your site.`

**Comment by zcwilt** [comment]
Sorry @torvista for taking a while to get back on this.

The only reason is that the creation of those files is automated, with some minor manual tweaking in some places. There is no other reason why those entries could not be tweaked further.

**Pull request merged**